redis/src/main/scala/zio/redis/RedisClusterExecutorLive.scala

```diff
@@ -0,0 +1,165 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.redis
+
+import zio._
+import zio.redis.RedisClusterExecutorLive._
+import zio.redis.api.Cluster.AskingCommand
+import zio.redis.codec.StringUtf8Codec
+import zio.redis.options.Cluster._
+import zio.schema.codec.Codec
+
+import java.io.IOException
+
+private[redis] final case class RedisClusterExecutorLive(
+  clusterConnectionRef: Ref.Synchronized[ClusterConnection],
+  scope: Scope.Closeable
+) extends RedisExecutor {
+
+  def execute(command: Chunk[RespValue.BulkString]): IO[RedisError, RespValue] = {
+
+    def executeAsk(executorZIO: IO[RedisError.IOError, RedisExecutor], asking: Boolean = false) =
+      for {
+        executor <- executorZIO
+        _        <- ZIO.when(asking)(executor.execute(AskingCommand.resp((), StringUtf8Codec)))
+        res      <- executor.execute(command)
+      } yield res
+
+    def executeSafe(keySlot: Slot) = {
+      val recover = executeAsk(getExecutor(keySlot)).flatMap {
+        case e: RespValue.Error => ZIO.fail(e.toRedisError)
+        case success            => ZIO.succeed(success)
+      }.catchSome {
+        case e: RedisError.Ask   => executeAsk(getExecutor(e.address), asking = true)
+        case _: RedisError.Moved => refreshConnect() *> executeAsk(getExecutor(keySlot))
+      }
+      recover.retry(Scheduler)
+    }
+
+    for {
+      key    <- ZIO.attempt(command(1)).orElseFail(CusterKeyError)
+      keySlot = Slot((key.asCRC16 % SlotsAmount).toLong)
+      result <- executeSafe(keySlot)
+    } yield result
+  }
+
+  private def getExecutor(slot: Slot): UIO[RedisExecutor] =
+    clusterConnectionRef.get.map(cc => cc.executor(slot))
+
+  // TODO introduce max connection amount
+  private def getExecutor(address: RedisUri) =
+    clusterConnectionRef.modifyZIO { cc =>
+      val executorOpt       = cc.executors.get(address).map(es => (es.executor, cc))
+      val enrichedClusterIO = scope.extend(connectToNode(address)).map(es => (es.executor, cc.addExecutor(address, es)))
+      ZIO.fromOption(executorOpt).catchAll(_ => enrichedClusterIO)
+    }
+
+  private def refreshConnect(): ZIO[Any, RedisError, Unit] =
+    clusterConnectionRef.updateZIO { connection =>
+      val addresses = connection.partitions.flatMap(_.addresses)
+      for {
+        cluster <- scope.extend(initConnectToCluster(addresses))
+        _       <- ZIO.foreachParDiscard(connection.executors) { case (_, es) => es.scope.close(Exit.unit) }
+      } yield cluster
+    }
+
+  // TODO configurable
+  private val Scheduler: Schedule[Any, Throwable, (zio.Duration, Long, Throwable)] =
+    Schedule.exponential(100.millis, 1.5) && Schedule.recurs(5) && Schedule.recurWhile[Throwable] {
+      case _: RedisError.IOError | _: RedisError.ClusterRedisError => true
+      case _                                                       => false
+    }
```

Comment:
It can be moved to the companion object. Also, it's not a scheduler, rather a retry policy.

Reply:
I want to do it configurable, if i do it, should i keep it here?

Reply:
Sure, if you plan to make it flexible keep it here.

Reply:
Done